File: ios/HybridNitroNetworkInfo.swift
//
//  HybridNitroNetworkInfo.swift
//  NitroNetworkInfo
//
//  Created by Patrick Kabwe on 04/05/2025.
//

import Foundation

typealias NetworkInfoListener = (NitroNetworkStatusInfo) -> Void

class HybridNitroNetworkInfo: HybridNitroNetworkInfoSpec {
    private var nitrogenNetworkInfoImpl = NitroNetworkInfoImpl()
    private var networkInfoStatusListener: ((NitroNetworkStatusInfo) -> Void)? = nil

    var isConnected: Bool {
        get { nitrogenNetworkInfoImpl.getIsConnected() }
    }
    
    var connectionType: ConnectionType {
        get { nitrogenNetworkInfoImpl.getConnectionType() }
    }
    
    func addListener(listener: @escaping (NitroNetworkStatusInfo) -> Void) throws -> () -> Void {
        networkInfoStatusListener = listener
        nitrogenNetworkInfoImpl.start()
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            
            self.networkInfoStatusListener?(
                NitroNetworkStatusInfo(
                    isConnected: self.isConnected,
                    connectionType: self.connectionType
                )
            )
            
            self.nitrogenNetworkInfoImpl.delegate = self
        }

        return { [weak self] in
            self?.networkInfoStatusListener = nil
            self?.nitrogenNetworkInfoImpl.stop()
        }
    }
    
    deinit {
        self.nitrogenNetworkInfoImpl.stop()
    }
}

extension HybridNitroNetworkInfo: NitroNetworkInfoDelegate {
    func nitroNetworkInfoDidChange(_ nitroNetworkInfo: NitroNetworkStatusInfo) {
        self.networkInfoStatusListener?(nitroNetworkInfo)
    }
}

File: android/src/main/java/com/nitronetworkinfo/HybridNitroNetworkInfo.kt
package com.nitronetworkinfo

import com.margelo.nitro.NitroModules
import com.margelo.nitro.nitronetworkinfo.ConnectionType
import com.margelo.nitro.nitronetworkinfo.HybridNitroNetworkInfoSpec
import com.margelo.nitro.nitronetworkinfo.NitroNetworkStatusInfo


class HybridNitroNetworkInfo: HybridNitroNetworkInfoSpec(), NetworkInfoDelegate {
    private val context = NitroModules.applicationContext ?: throw Exception("Context is null")
    private val nitroNetworkInfoImpl = NitroNetworkInfoImpl(context, this)
    private var networkInfoStatusListener: ((NitroNetworkStatusInfo) -> Unit)? = null

    override val isConnected: Boolean
        get() = nitroNetworkInfoImpl.getIsConnection()
    override val connectionType: ConnectionType
        get() = nitroNetworkInfoImpl.getConnectionType()

    override fun addListener(listener: (NitroNetworkStatusInfo) -> Unit): () -> Unit {
        networkInfoStatusListener = listener
        nitroNetworkInfoImpl.start()
        listener(NitroNetworkStatusInfo(isConnected, connectionType))
        return {
            networkInfoStatusListener = null
            nitroNetworkInfoImpl.stop()
        }
    }

    override fun onNetworkInfoChanged(info: NitroNetworkStatusInfo) {
        networkInfoStatusListener?.invoke(info)
    }
}
